garage/tf/algos/dqn.py

```diff
@@ -0,0 +1,206 @@
+        self.double_q = double_q
+
+        self.episode_rewards = []
+        self.episode_qf_losses = []
+
+        super().__init__(
+            env_spec=env_spec,
+            replay_buffer=replay_buffer,
+            max_path_length=max_path_length,
+            discount=discount,
+            **kwargs)
+
+    @overrides
+    def init_opt(self):
+        """
+        Initialize the networks and Ops.
+
+        Assume discrete space for dqn, so action dimension
+        will always be action_space.n
+        """
+        action_dim = self.env_spec.action_space.n
+
+        # build q networks
+        with tf.name_scope(self.name, 'DQN'):
+            self.action_t_ph = tf.placeholder(tf.int32, None, name='action')
+            self.reward_t_ph = tf.placeholder(tf.float32, None, name='reward')
+            self.done_t_ph = tf.placeholder(tf.float32, None, name='done')
+
+            # clone a target q-function
+            self.target_qf = self.qf.clone('target_qf')
+
+            with tf.name_scope('update_ops'):
+                target_update_op = tensor_utils.get_target_ops(
+                    self.qf.get_global_vars(),
+                    self.target_qf.get_global_vars())
+
+            self._qf_update_ops = tensor_utils.compile_function(
+                inputs=[], outputs=target_update_op)
+
+            with tf.name_scope('td_error'):
+                # Q-value of the selected action
+                action = tf.one_hot(self.action_t_ph, action_dim)
+                q_selected = tf.reduce_sum(
+                    self.qf.q_vals * action,  # yapf: disable
+                    axis=1)
+
+                # r + Q'(s', argmax_a(Q(s', _)) - Q(s, a)
+                if self.double_q:
+                    target_qval_with_online_q = self.qf.get_qval_sym(
+                        self.target_qf.input, self.qf.name)
+                    future_best_q_val_action = tf.argmax(
+                        target_qval_with_online_q, 1)
+                    future_best_q_val = tf.reduce_sum(
+                        self.target_qf.q_vals * tf.one_hot(
+                            future_best_q_val_action, action_dim),
+                        axis=1)
+                else:
+                    # r + max_a(Q'(s', _)) - Q(s, a)
+                    future_best_q_val = tf.reduce_max(
+                        self.target_qf.q_vals, axis=1)
+
+                q_best_masked = (1.0 - self.done_t_ph) * future_best_q_val
+                # if done, it's just reward
+                # else reward + discount * future_best_q_val
+                target_q_values = self.reward_t_ph + self.discount * q_best_masked  # noqa: E501
+
+                td_error = q_selected - tf.stop_gradient(target_q_values)
+                loss = huber_loss(td_error)
+                self._loss = tf.reduce_mean(loss)
+
+            with tf.name_scope('optimize_ops'):
+                optimizer = self.qf_optimizer(self.qf_lr)
+                if self.grad_norm_clipping is not None:
```

Comment:
this mode is not tested at all. please test it or remove.

Reply:
benchmarks work fine without clipping gradient. Maybe we can just remove it.

Reply:
didn't you say it's required to compare to openai/baselines?

Reply:
openai/baselines use gradient clipping. I tried and benchmarks work comparable with them without gradient clipping. If we want to use exactly the same setting as theirs, we can also do gradient clipping.

Reply:
let's leave it for now but make a comment about why it's there. i'd prefer if you added a test which covered it.

Reply:
I think we can keep it? It is recommended to do gradient clipping with DQN in practice.

Reply:
Sure. I also added a test for gradient clipping.